File: Problems 0500-0599/502. IPO/solution.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        # heap that contains all projects with enough capital
        # to start
        n = len(profits)
        projects = sorted(zip(profits, capital), key=lambda x: x[1])
        
        i = 0
        capital = w
        projectsToStart = []
        for _ in range(k):
            while i < n and projects[i][1] <= capital:
                heappush(projectsToStart, -projects[i][0])
                i += 1
            
            if not projectsToStart:
                break
            
            profit = -heappop(projectsToStart)
            capital += profit
        
        return capital

File: Problems 0500-0599/502. IPO/solution.cpp
class Solution {
    public:
        int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
            int n = profits.size();
            vector<pair<int, int>> projects;
            for (int i = 0; i < n; i++) {
                projects.emplace_back(profits[i], capital[i]);
            }
    
            sort(projects.begin(), projects.end(), [] (const pair<int, int>& a, const pair<int, int>& b) {
                return a.second < b.second;
            });
    
            priority_queue<int, vector<int>> maxHeap;
    
            int i = 0;
            while (k--) {
                while (i < n && projects[i].second <= w) {
                    maxHeap.emplace(projects[i].first);
                    i++;
                }
    
                if (maxHeap.empty()) break;
    
                int profit = maxHeap.top();
                maxHeap.pop();
                w += profit;
            }
            return w;
        }
    };